// doc-prsr/dist/typedoc.parser/parsers/overview.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var typedoc_parser_options_1 = require("../typedoc.parser.options");
var overview_1 = require("../../model/class/overview");
var examples_parser_1 = require("./examples.parser");
var OverviewParser = /** @class */ (function () {
    function OverviewParser() {
        this.exampleParser = new examples_parser_1.ExamplesParser();
    }
    OverviewParser.prototype.getOverview = function (obj) {
        var _this = this;
        var _a;
        var tags = obj[typedoc_parser_options_1.CO.comment][typedoc_parser_options_1.CO.tags];
        var content = tags && tags.map(function (example) { return _this.exampleParser.getNodes(example); })
            .filter(function (example) { return example && example[0].type !== overview_1.OverviewNodeType.ADDITIONAL_EXAMPLE; });
        return (_a = [
            this.getShortDescription(obj),
            this.getDescription(obj),
        ]).concat.apply(_a, content).filter(function (tag) { return tag && tag.content; });
    };
    OverviewParser.prototype.getDescription = function (obj) {
        return {
            type: overview_1.OverviewNodeType.TEXT,
            content: obj && obj[typedoc_parser_options_1.CO.comment] && obj[typedoc_parser_options_1.CO.comment]['text'],
        };
    };
    OverviewParser.prototype.getShortDescription = function (obj) {
        return {
            type: overview_1.OverviewNodeType.TEXT,
            content: obj && obj[typedoc_parser_options_1.CO.comment] && obj[typedoc_parser_options_1.CO.comment]['shortText'],
        };
    };
    return OverviewParser;
}());
exports.OverviewParser = OverviewParser;
