// doc-prsr/dist/doc-js.parser/doc-js.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = require("../model");
var doc_js_parser_options_1 = require("./doc-js.parser.options");
var parsers_1 = require("./parsers");
var DocJsParser = /** @class */ (function () {
    function DocJsParser() {
        this.styles = new parsers_1.StylesParser();
        this.props = new parsers_1.PropertiesParser();
        this.examples = new parsers_1.ExamplesParser();
        this.methods = new parsers_1.MethodsParser();
        this.common = new parsers_1.Common();
    }
    DocJsParser.prototype.parse = function (json, metadata) {
        this.saveJSON(json);
        return new model_1.Model(this.getClasses(this.json), metadata);
    };
    DocJsParser.prototype.saveJSON = function (json) {
        this.json = json;
    };
    DocJsParser.prototype.getClasses = function (json) {
        var _this = this;
        return json
            .filter(function (item) { return _this.isClass(item); })
            .map(function (item) { return _this.parseClass(item); });
    };
    DocJsParser.prototype.parseClass = function (obj) {
        return new model_1.Class({
            kind: this.common.getKind(obj),
            platform: null,
            examples: this.examples.getExamples(obj),
            props: this.props.getProps(obj),
            methods: this.methods.getMethods(obj),
            name: this.common.getName(obj),
            shortDescription: this.common.getShortDescription(obj),
            description: this.common.getDescription(obj),
            styles: this.styles.getStyles(obj),
            // TODO create docs-js examples parser
            liveExamples: [],
            // TODO create docs-js overview parser
            overview: []
        });
    };
    DocJsParser.prototype.isClass = function (obj) {
        return obj[doc_js_parser_options_1.CO.classKind] === 'class';
    };
    return DocJsParser;
}());
exports.DocJsParser = DocJsParser;
