// doc-prsr/dist/model/class/example.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/** @deprecated */
var Sample = /** @class */ (function () {
    function Sample(options) {
        this.shortDescription = options.shortDescription;
        this.description = options.description;
        this.code = options.code;
    }
    return Sample;
}());
exports.Sample = Sample;
var Example = /** @class */ (function () {
    function Example() {
    }
    return Example;
}());
exports.Example = Example;
var InlineExample = /** @class */ (function (_super) {
    __extends(InlineExample, _super);
    function InlineExample() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return InlineExample;
}(Example));
exports.InlineExample = InlineExample;
