// doc-prsr/dist/doc-js.parser/parsers/examples.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = require("../../model");
var doc_js_parser_options_1 = require("../doc-js.parser.options");
var ExamplesParser = /** @class */ (function () {
    function ExamplesParser() {
    }
    ExamplesParser.prototype.getExamples = function (obj) {
        var _this = this;
        if (obj[doc_js_parser_options_1.CO.examples].length) {
            return obj[doc_js_parser_options_1.CO.examples].map(function (item) { return _this.parseExample(item); });
        }
        else {
            return [];
        }
    };
    ExamplesParser.prototype.parseExample = function (obj) {
        return new model_1.Sample({
            code: this.getCode(obj),
            description: this.getDescription(obj),
            shortDescription: this.getShortDescription(obj)
        });
    };
    ExamplesParser.prototype.splitExample = function (example) {
        if (example[doc_js_parser_options_1.CO.description]) {
            return example[doc_js_parser_options_1.CO.description].split(/```/g);
        }
        else {
            return [];
        }
    };
    ExamplesParser.prototype.getCode = function (example) {
        var splittedExample = this.splitExample(example);
        if (splittedExample.length !== 0) {
            return splittedExample[1];
        }
        else {
            return '';
        }
    };
    ExamplesParser.prototype.getDescription = function (example) {
        var splittedExample = this.splitExample(example);
        var shortDescription = splittedExample[0].split('\r\n')[0];
        if (splittedExample.length !== 0) {
            return splittedExample[0].replace(shortDescription, '');
        }
        else {
            return '';
        }
    };
    ExamplesParser.prototype.getShortDescription = function (example) {
        var splittedExample = this.splitExample(example);
        if (splittedExample.length !== 0) {
            return splittedExample[0].split('\r\n')[0];
        }
        else {
            return '';
        }
    };
    return ExamplesParser;
}());
exports.ExamplesParser = ExamplesParser;
