// doc-prsr/dist/doc-js.parser/parsers/params.parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var model_1 = require("../../model");
var doc_js_parser_options_1 = require("../doc-js.parser.options");
var common_1 = require("./common");
var ParamsParser = /** @class */ (function () {
    function ParamsParser() {
        this.common = new common_1.Common();
    }
    ParamsParser.prototype.getParams = function (obj) {
        var _this = this;
        if (obj[doc_js_parser_options_1.CO.params] && obj[doc_js_parser_options_1.CO.params].length) {
            return obj[doc_js_parser_options_1.CO.params].map(function (item) { return _this.parseParam(item); });
        }
        else {
            return [];
        }
    };
    ParamsParser.prototype.parseParam = function (obj) {
        return new model_1.Param({
            name: this.common.getName(obj),
            type: this.getType(obj),
            required: null,
            shortDescription: this.common.getShortDescription(obj),
            description: this.common.getDescription(obj)
        });
    };
    ParamsParser.prototype.getType = function (param) {
        return param[doc_js_parser_options_1.CO.paramType] ? param[doc_js_parser_options_1.CO.paramType][doc_js_parser_options_1.CO.name] : '';
    };
    return ParamsParser;
}());
exports.ParamsParser = ParamsParser;
